### DISCUSSION TITLE
DI and Creating Connection

### DISCUSSION THREAD

**Discussion by TechyChap** [abuzuhri/Amazon-SP-API-CSharp#144] [General]
Any suggestions about how to go about using .Net Core DI to create the AmazonConnection ?

Am thinking a connection factory approach so I can connect to different marketplaces.

**Comment by abuzuhri** [comment]
I'm not sour what the best for you but for peoples have multi marketplaces we have **RefNumber** can filled and later get object by this number

```Csharp
AmazonConnection amazonConnection = new AmazonConnection(new AmazonCredential()
            {
                .
                .
            },
            RefNumber:"ABCD...."
            );
```

**Comment by TechyChap** [comment]
A related question. Should the connection be short lived or long/singleton?

**Reply by abuzuhri** [reply]
Connection should be for long time if you going to use  api in heavy use  , all information about rate limit need the connection to be live to do the true calculations for next request , and save access token with expired time .

If you lite use then no need keep it for long time

**Reply by TechyChap** [reply]
Great. That's enough for me to work with. Thanks

**Reply by rcalv002** [reply]
> Great. That's enough for me to work with. Thanks

Hey @TechyChap ,

Were you able to get this working reliably as a singleton?

**Reply by TechyChap** [reply]
Not seen any issues yet. we are using .Net Core and the connection is currently defined as:

 services.AddSingleton<AmazonConnection>(c => new AmazonConnection(new AmazonCredential()
            {
                AccessKey = settings.AccessKey,
                SecretKey = settings.SecretKey,
                RoleArn = settings.RoleArn,
                ClientId = settings.ClientId,
                ClientSecret = settings.ClientSecret,
                RefreshToken = settings.RefreshToken,
                MarketPlace = MarketPlace.UnitedKingdom, 
                Environment = settings.UseSandbox ? Constants.Environments.Sandbox : Constants.Environments.Production
            }));
            
            
We've not implemented the multiple marketplaces logic yet.

**Reply by TechyChap** [reply]
We ended up with a simple wrapper around the SP-API to allow unit testing without the needing to mock out the whole connection.

```
public class AmazonListingService : IAmazonListingService
    {
        private readonly AmazonConnection _connection;

        public AmazonListingService(AmazonConnection connection)
        {
            _connection = connection;
        }

        public async Task<ListingsItemSubmissionResponse> PutListingsItemAsync(ParameterPutListingItem listing)
        {
            return await _connection.ListingsItem.PutListingsItemAsync(listing);
        }
    }
```
